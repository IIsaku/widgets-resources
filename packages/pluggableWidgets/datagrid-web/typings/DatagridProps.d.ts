/**
 * This file was generated from Datagrid.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListExpressionValue, ListWidgetValue } from "mendix";

export type HidableEnum = "yes" | "hidden" | "no";

export type ShowContentAsEnum = "attribute" | "dynamicText" | "customContent";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    header?: DynamicValue<string>;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    filter?: ReactNode;
    showContentAs: ShowContentAsEnum;
    attribute?: ListAttributeValue<string | BigJs.Big | boolean | Date>;
    content?: ListWidgetValue;
    dynamicText?: ListExpressionValue<string>;
    width: WidthEnum;
    size: number;
    alignment: AlignmentEnum;
    columnClass?: ListExpressionValue<string>;
}

export type PaginationEnum = "buttons" | "virtualScrolling";

export type PagingPositionEnum = "bottom" | "top";

export interface ColumnsPreviewType {
    header: string;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    filter: { widgetCount: number; renderer: ComponentType };
    showContentAs: ShowContentAsEnum;
    attribute: string;
    content: { widgetCount: number; renderer: ComponentType };
    dynamicText: string;
    width: WidthEnum;
    size: number | null;
    alignment: AlignmentEnum;
    columnClass: string;
}

export interface DatagridContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    datasource: ListValue;
    columns: ColumnsType[];
    showEmptyPlaceholder: boolean;
    emptyPlaceholder?: ReactNode;
    onClick?: ListActionValue;
    rowClass?: ListExpressionValue<string>;
    columnsFilterable: boolean;
    pageSize: number;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute?: EditableValue<string>;
    onConfigurationChange?: ActionValue;
}

export interface DatagridPreviewProps {
    class: string;
    style: string;
    datasource: {} | null;
    columns: ColumnsPreviewType[];
    showEmptyPlaceholder: boolean;
    emptyPlaceholder: { widgetCount: number; renderer: ComponentType };
    onClick: {} | null;
    rowClass: string;
    columnsFilterable: boolean;
    pageSize: number | null;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute: string;
    onConfigurationChange: {} | null;
}
