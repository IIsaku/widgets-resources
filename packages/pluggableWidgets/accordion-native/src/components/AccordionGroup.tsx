import { createElement, ReactElement } from "react";
import { View, Pressable, Text } from "react-native";
import { DynamicValue, NativeIcon } from "mendix";

import { GroupIcon } from "./GroupIcon";
import { AnimatedCollapsibleView } from "./CollapsibleView";
import { GroupsType, IconEnum } from "../../typings/AccordionProps";
import { AccordionGroupStyle } from "../ui/Styles";

export interface AccordionGroupProps {
    index: number;
    collapsible: boolean;
    icon: IconEnum;
    iconCollapsed: DynamicValue<NativeIcon> | undefined;
    iconExpanded: DynamicValue<NativeIcon> | undefined;
    group: GroupsType;
    isExpanded: boolean;
    onPressGroupHeader: (group: GroupsType, index: number) => void;
    visible: boolean;
    style: AccordionGroupStyle;
}

export function AccordionGroup({
    index,
    collapsible,
    icon,
    iconCollapsed,
    iconExpanded,
    group,
    isExpanded,
    onPressGroupHeader,
    visible,
    style
}: AccordionGroupProps): ReactElement | null {
    return visible ? (
        <View style={style.container}>
            <Pressable
                style={[style.header.container, icon === "left" && { flexDirection: "row-reverse" }]}
                onPress={collapsible ? () => onPressGroupHeader(group, index) : null}
            >
                {group.headerRenderMode === "text" ? (
                    <Text style={[style.header[group.headerTextRenderMode], { flex: 1 }]}>
                        {group.headerText.value}
                    </Text>
                ) : (
                    <View style={{ flex: 1 }}>{group.headerContent}</View>
                )}
                {icon !== "no" && collapsible ? (
                    <GroupIcon
                        isExpanded={isExpanded}
                        iconCollapsed={iconCollapsed}
                        iconExpanded={iconExpanded}
                        style={style.header.icon}
                    />
                ) : null}
            </Pressable>
            <AnimatedCollapsibleView isExpanded={isExpanded} style={style.content}>
                {group.content}
            </AnimatedCollapsibleView>
        </View>
    ) : null;
}
