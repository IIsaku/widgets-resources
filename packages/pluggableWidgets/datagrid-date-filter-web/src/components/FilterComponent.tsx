import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "@mendix/piw-utils-internal";

import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";

import DatePickerComponent from "react-datepicker";
import { DatePicker } from "./DatePicker";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    defaultValue?: Date;
    dateFormat?: string;
    locale?: string;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    updateFilters?: (value: Date | null, type: DefaultFilterEnum) => void;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Date | null>(null);
    const pickerRef = useRef<DatePickerComponent | null>(null);

    useEffect(() => {
        if (props.defaultValue) {
            setValue(props.defaultValue);
        }
    }, [props.defaultValue]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const focusInput = useCallback(() => {
        if (pickerRef.current) {
            pickerRef.current.setFocus();
        }
    }, [pickerRef.current]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    defaultFilter={props.defaultFilter}
                    name={props.name}
                    onChange={useCallback(
                        type => {
                            setType(type);
                            focusInput();
                        },
                        [focusInput]
                    )}
                    options={
                        [
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                dateFormat={props.dateFormat}
                locale={props.locale}
                name={props.name}
                placeholder={props.placeholder}
                ref={pickerRef}
                screenReaderInputCaption={props.screenReaderInputCaption}
                setValue={setValue}
                value={value}
            />
        </div>
    );
}
