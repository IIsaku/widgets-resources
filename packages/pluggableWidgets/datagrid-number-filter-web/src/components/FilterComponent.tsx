import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { debounce, FilterSelector } from "@mendix/piw-utils-internal";

import { DefaultFilterEnum } from "../../typings/DatagridNumberFilterProps";
import { Big } from "big.js";
import classNames from "classnames";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    updateFilters?: (value: Big | undefined, type: DefaultFilterEnum) => void;
    value?: Big;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Big | undefined>(undefined);
    const [valueInput, setValueInput] = useState<string | undefined>(undefined);
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value.toString());
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const onChange = useCallback(
        debounce((value?: Big) => setValue(value), props.delay),
        [props.delay]
    );

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    name={props.name}
                    defaultFilter={props.defaultFilter}
                    onChange={useCallback(
                        type => {
                            setType(type);
                            focusInput();
                        },
                        [focusInput]
                    )}
                    options={
                        [
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                onChange={e => {
                    const value = e.target.value;
                    if (value && !isNaN(Number(value))) {
                        setValueInput(value);
                        onChange(new Big(Number(value)));
                    } else {
                        setValueInput(value);
                        onChange(undefined);
                    }
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="number"
                value={valueInput}
            />
        </div>
    );
}
