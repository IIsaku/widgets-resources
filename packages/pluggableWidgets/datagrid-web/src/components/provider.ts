import { createContext } from "react";
import { FilterContextValue } from "@mendix/piw-utils-internal";

export const FilterContext = createContext((undefined as any) as FilterContextValue);

(window as any)["com.mendix.widgets.web.datagrid.filterContext"] = FilterContext;
