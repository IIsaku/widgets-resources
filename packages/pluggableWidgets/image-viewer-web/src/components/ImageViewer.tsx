import { createElement, CSSProperties, ReactElement } from "react";
import classNames from "classnames";
import { HeightUnitEnum, WidthUnitEnum } from "../../typings/ImageViewerProps";

import "../ui/ImageViewer.scss";

function getStyle(value: string | number, type: WidthUnitEnum | HeightUnitEnum): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}
export interface ImageViewerWrapperProps {
    className?: string;
    responsive: boolean;
    hasImage: boolean;
    children: ReactElement<ImageViewerGlyphicon | ImageViewerImage>;
}

interface ImageViewerContent {
    style?: CSSProperties;
}

function Wrapper(props: ImageViewerWrapperProps): ReactElement {
    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.hasImage }
            )}
        >
            {props.children}
        </div>
    );
}

export interface ImageViewerGlyphicon extends ImageViewerContent {
    icon: string | undefined;
    size: number;
}

function Glyphicon(props: ImageViewerGlyphicon): ReactElement {
    return (
        <span className={classNames("glyphicon", props.icon)} style={{ ...props.style, fontSize: `${props.size}px` }} />
    );
}

export interface ImageViewerImage extends ImageViewerContent {
    image: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
}

function Image(props: ImageViewerImage): ReactElement {
    return (
        <img
            src={props.image}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit)
            }}
        />
    );
}

export const ImageViewer = {
    Wrapper,
    Glyphicon,
    Image
};
