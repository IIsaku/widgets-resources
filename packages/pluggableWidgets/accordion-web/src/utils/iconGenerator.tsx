import { createElement, ReactElement, useCallback } from "react";

import { Icon, IconProps } from "../components/Icon";

import { AccordionContainerProps } from "../../typings/AccordionProps";

export function useIconGenerator(
    advancedMode: AccordionContainerProps["advancedMode"],
    animateIcon: AccordionContainerProps["animateIcon"],
    icon: Omit<IconProps, "animate">,
    expandIcon: Omit<IconProps, "animate">,
    collapseIcon: Omit<IconProps, "animate">
): (collapsed: boolean) => ReactElement {
    return useCallback(
        (collapsed: boolean): ReactElement => {
            if (!advancedMode) {
                return <Icon animate={animateIcon} />;
            } else {
                if (animateIcon) {
                    return <Icon {...icon} animate={animateIcon} />;
                } else {
                    return collapsed ? (
                        <Icon {...expandIcon} animate={animateIcon} />
                    ) : (
                        <Icon {...collapseIcon} animate={animateIcon} />
                    );
                }
            }
        },
        [advancedMode, animateIcon, icon, expandIcon, collapseIcon]
    );
}
