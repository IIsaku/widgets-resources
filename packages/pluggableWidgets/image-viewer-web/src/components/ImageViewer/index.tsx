import { cloneElement, createElement, CSSProperties, FunctionComponent, useCallback } from "react";
import { HeightUnitEnum, WidthUnitEnum, OnClickTypeEnum, DisplayAsEnum } from "../../../typings/ImageViewerProps";
import { useLightboxState } from "../../utils/lightboxState";
import { ImageViewerUi, ImageViewerContentProps } from "./ui";
import { Lightbox, LightboxProps } from "../Lightbox";

import "../../ui/ImageViewer.scss";

export type ImageViewerImageProps = {
    type: "image" | "icon";
    image: string | undefined;
};

export interface ImageViewerProps extends ImageViewerImageProps {
    class: string;
    style?: CSSProperties;
    widthUnit: WidthUnitEnum;
    width: number;
    heightUnit: HeightUnitEnum;
    height: number;
    iconSize: number;
    responsive: boolean;
    onClickType: OnClickTypeEnum;
    onClick?: () => void;
    altText?: string;
    displayAs: DisplayAsEnum;
    previewMode?: boolean;
}

function processImageLink(imageLink: string | undefined, displayAs: DisplayAsEnum): string | undefined {
    if (!imageLink || displayAs === "fullImage") {
        return imageLink;
    }
    const url = new URL(imageLink);
    url.searchParams.append("thumb", "true");
    return url.href;
}

export const ImageViewer: FunctionComponent<ImageViewerProps> = ({
    class: className,
    style,
    widthUnit,
    width,
    heightUnit,
    height,
    iconSize,
    responsive,
    onClickType,
    onClick,
    type,
    image,
    altText,
    displayAs,
    previewMode
}) => {
    const { lightboxIsOpen, openLightbox, closeLightbox } = useLightboxState();

    const onCloseLightbox = useCallback<LightboxProps["onClose"]>(
        event => {
            event?.stopPropagation();
            closeLightbox();
        },
        [closeLightbox]
    );

    const onImageClick = useCallback<Exclude<ImageViewerContentProps["onClick"], undefined>>(
        event => {
            event.stopPropagation();
            if (onClickType === "action") {
                onClick?.();
            } else if (onClickType === "enlarge") {
                openLightbox();
            }
        },
        [onClick, onClickType, openLightbox]
    );

    const hasClickHandler = (onClickType === "action" && onClick) || onClickType === "enlarge";
    const sharedContentProps: ImageViewerContentProps = {
        style,
        onClick: hasClickHandler ? onImageClick : undefined,
        altText
    };

    const content =
        type === "image" ? (
            <ImageViewerUi.Image
                image={processImageLink(image, displayAs)}
                height={height}
                heightUnit={heightUnit}
                width={width}
                widthUnit={widthUnit}
                {...sharedContentProps}
            />
        ) : (
            <ImageViewerUi.Glyphicon icon={image} size={iconSize} {...sharedContentProps} />
        );

    return (
        <ImageViewerUi.Wrapper
            className={className}
            responsive={responsive}
            hasImage={image !== undefined && image.length > 0}
        >
            {content}
            {!previewMode && lightboxIsOpen && (
                <Lightbox isOpen={lightboxIsOpen} onClose={onCloseLightbox}>
                    {type === "image" ? cloneElement(content, { image, onClick: undefined }) : content}
                </Lightbox>
            )}
        </ImageViewerUi.Wrapper>
    );
};
