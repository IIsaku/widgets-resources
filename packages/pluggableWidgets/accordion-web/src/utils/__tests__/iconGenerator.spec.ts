import { renderHook, RenderHookResult } from "@testing-library/react-hooks";
import { AccordionContainerProps } from "../../../typings/AccordionProps";
import { IconProps } from "../../components/Icon";
import { useIconGenerator } from "../iconGenerator";
import { ReactElement } from "react";

interface HookProps {
    advancedMode: AccordionContainerProps["advancedMode"];
    animateIcon: AccordionContainerProps["animateIcon"];
    icon: Omit<IconProps, "animate">;
    expandIcon: Omit<IconProps, "animate">;
    collapseIcon: Omit<IconProps, "animate">;
}

describe("useIconGenerator", () => {
    const defaultHookProps: HookProps = {
        advancedMode: false,
        animateIcon: true,
        icon: { data: { type: "glyph", iconClass: "icon-class" }, loading: false },
        expandIcon: { data: { type: "glyph", iconClass: "expand-icon-class" }, loading: false },
        collapseIcon: { data: { type: "glyph", iconClass: "collapse-icon-class" }, loading: false }
    };

    const initUseIconGenerator = (
        initialProps = defaultHookProps
    ): RenderHookResult<HookProps, (collapsed: boolean) => ReactElement> =>
        renderHook(
            (props: {
                advancedMode: AccordionContainerProps["advancedMode"];
                animateIcon: AccordionContainerProps["animateIcon"];
                icon: Omit<IconProps, "animate">;
                expandIcon: Omit<IconProps, "animate">;
                collapseIcon: Omit<IconProps, "animate">;
            }) =>
                useIconGenerator(
                    props.advancedMode,
                    props.animateIcon,
                    props.icon,
                    props.expandIcon,
                    props.collapseIcon
                ),
            {
                initialProps
            }
        );

    it("returns the same function when the arguments are the same", () => {
        const renderedHook = initUseIconGenerator();
        const firstFunction = renderedHook.result.current;

        renderedHook.rerender();
        expect(renderedHook.result.current).toBe(firstFunction);
    });
    it("returns a new function when the arguments have changed", () => {
        const renderedHook = initUseIconGenerator();
        const firstFunction = renderedHook.result.current;

        renderedHook.rerender({ ...defaultHookProps, animateIcon: false });
        expect(renderedHook.result.current).not.toBe(firstFunction);
    });

    describe("returned function", () => {
        it("returns the default icon when not in advanced mode", () => {
            const renderedHook = initUseIconGenerator({ ...defaultHookProps });
            expect(renderedHook.result.current(true)).toMatchSnapshot();
        });

        describe("with advanced mode", () => {
            it("returns the configured icon when icon is animated", () => {
                const renderedHook = initUseIconGenerator({ ...defaultHookProps, advancedMode: true });
                expect(renderedHook.result.current(true)).toMatchSnapshot();
            });
            it("returns the expand icon when collapsed", () => {
                const renderedHook = initUseIconGenerator({
                    ...defaultHookProps,
                    advancedMode: true,
                    animateIcon: false
                });
                expect(renderedHook.result.current(true)).toMatchSnapshot();
            });
            it("returns the collapse icon when expanded", () => {
                const renderedHook = initUseIconGenerator({
                    ...defaultHookProps,
                    advancedMode: true,
                    animateIcon: false
                });
                expect(renderedHook.result.current(false)).toMatchSnapshot();
            });
        });
    });
});
