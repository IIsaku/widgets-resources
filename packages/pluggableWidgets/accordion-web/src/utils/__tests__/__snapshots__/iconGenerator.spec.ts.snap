// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIconGenerator returned function returns the default icon when not in advanced mode 1`] = `
<Icon
  animate={true}
/>
`;

exports[`useIconGenerator returned function with advanced mode returns the collapse icon when expanded 1`] = `
<Icon
  animate={false}
  data={
    Object {
      "iconClass": "collapse-icon-class",
      "type": "glyph",
    }
  }
  loading={false}
/>
`;

exports[`useIconGenerator returned function with advanced mode returns the configured icon when icon is animated 1`] = `
<Icon
  animate={true}
  data={
    Object {
      "iconClass": "icon-class",
      "type": "glyph",
    }
  }
  loading={false}
/>
`;

exports[`useIconGenerator returned function with advanced mode returns the expand icon when collapsed 1`] = `
<Icon
  animate={false}
  data={
    Object {
      "iconClass": "expand-icon-class",
      "type": "glyph",
    }
  }
  loading={false}
/>
`;
